test/provider/AppDirectory.unittest.ts

```diff
@@ -0,0 +1,203 @@
+        this._blob = blob;
+        this._options = options;
+    }
+
+    public get ok() {
+        const {status} = this._options;
+        return status >= 200 && status <= 400;
+    }
+
+    public async json() {
+        return JSON.parse(this._blob);
+    }
+}
+
+const fakeAppsJSON = JSON.stringify(fakeApps);
+
+// Define localStorage global/window
+Object.defineProperty(global, 'localStorage', {
+    value: {
+        getItem: jest.fn(),
+        setItem: jest.fn()
+    },
+    writable: true
+});
+
+describe('AppDirectory Unit Tests', () => {
+    beforeEach(() => {
+        // Reset localStorage
+        global.localStorage = {
+            getItem: jest.fn().mockImplementation((key: string) => {
+                switch (key) {
+                    case StorageKeys.APPLICATIONS:
+                        return fakeAppsJSON;
+                    case StorageKeys.URL:
+                        return 'http://localhost:3923/provider/sample-app-directory.json';
+                    default:
+                        return null;
+                }
+            }),
+            setItem: jest.fn((key: string, value: string) => {})
+        };
+        global.fetch = jest.fn().mockImplementation(async (url: string) => {
+            return new Response(fakeAppsJSON, {status: 200});
+        });
+    });
+
+    describe('When directory is initialized', () => {
+        it('Loads directory and URL from the store', async () => {
+            const appDirectory = new AppDirectory();
+            expect(appDirectory['_directory']).toEqual(fakeApps);
+            expect(appDirectory['_URL']).toEqual('http://localhost:3923/provider/sample-app-directory.json');
+        });
+
+        it('Handles corrupt cached JSON in the store & resets it', () => {
+            global.localStorage.getItem.mockImplementation((key: string) => {
+                return StorageKeys.APPLICATIONS ? '__not_valid_json__' : 'http://localhost';
+            });
+
+            const appDirectory = new AppDirectory();
+            // Clear the corrupted data
+            expect(localStorage.getItem).toBeCalled();
+            expect(appDirectory['_directory']).toEqual([]);
+        });
+    });
+
+    describe('When querying the Directory', () => {
+        describe('Refreshing the directory', () => {
+            it('Fetches & updates the directory when URLs do not match', async () => {
+                const appDirectory = new AppDirectory();
+                const newApplications = [{...appTemplate, name: 'test-app'}];
+                const spyGetItem = jest.spyOn(global.localStorage, 'getItem');
+                const spyFetch = jest.spyOn(global, 'fetch').mockImplementation(async () => {
+                    return new Response(JSON.stringify(newApplications), {status: 200});
+                });
+                spyGetItem.mockImplementation(() => '__test_url__');
+                appDirectory['_URL'] = '*different_url*';
```

Comment:
In real world usage, what would be the right way to change what URL `AppDirectory` uses? Just want to check this is what we want to do here, rather than setting a new URL via localStorage?

Reply:
Exact details are still TBD, but at the high level will be using the config mechanisms from layouts. A JSON schema will define the available options, and the directory will query the config store for the URL.

Reply:
So are we content that this is a good way to change the URL for the purposes of this unit test?

Reply:
I think it's probably fine for now. I'd rather have a test which works now but needs updating than wait until we have everything perfect to test with.

Reply:
👍